### PR TITLE
Integration Test for SCIM2 Multi Attribute User Filter

### PR DESCRIPTION
**Pull request by sarubi**
Support "and" logical operator for SCIM2 filters such as EQ, SW, EW, CO. The user can query user details using multi-attribute search filters with pagination.
For further details,
[https://docs.google.com/document/d/1FopEfAE2fHBki9A-uw81-VsiG9FJkVXEk_ESADglNsA/edit?usp=sharing](url)

Depends on following PRs,
[1][https://github.com/wso2/carbon-kernel/pull/1817]()
[2]https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/112
[3]https://github.com/wso2/charon/pull/133

### CODE REVIEW

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java [hunk 1]
```diff
@@ -0,0 +1,378 @@
+package org.wso2.identity.integration.test.scim2;
```
Comment (thanujalk) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java]: Add license header

Comment (sarubi) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java]: Fixed

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+        EntityUtils.consume(response.getEntity());
+
+        userResourcePath = getUserPath() + "/" + userId;
+        HttpGet getRequest = new HttpGet(userResourcePath);
+        getRequest.addHeader(HttpHeaders.AUTHORIZATION, getAuthzHeader());
+        getRequest.addHeader(HttpHeaders.CONTENT_TYPE, "application/json");
+
+        response = client.execute(request);
+        assertEquals(response.getStatusLine().getStatusCode(), 404, "User " +
+                "has not been deleted successfully");
```
Comment (Abilashini) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java]: This line can be merged with the previous line. Seems like it does not exceed 120 characters.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+
+    private void deleteUser(String userId) throws IOException {
+
+        String userResourcePath = getUserPath() + "/" + userId;
+        HttpDelete request = new HttpDelete(userResourcePath);
+        request.addHeader(HttpHeaders.AUTHORIZATION, getAuthzHeader());
+        request.addHeader(HttpHeaders.CONTENT_TYPE, "application/json");
+
+        HttpResponse response = client.execute(request);
+        assertEquals(response.getStatusLine().getStatusCode(), 204, "User " +
```
Comment (Abilashini) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java]: Same applies here https://github.com/wso2/product-is/pull/3601/files#r317469532

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+
+        rootObject.put(SCIM2BaseTestCase.MEMBERS_ATTRIBUTE, members);
+
+        StringEntity entity = new StringEntity(rootObject.toString());
+        request.setEntity(entity);
+
+        HttpResponse response = client.execute(request);
+
+        assertEquals(response.getStatusLine().getStatusCode(), 201, "User " +
+                "has not been created successfully");
```
Comment (Abilashini) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java]: https://github.com/wso2/product-is/pull/3601/files#r317469532

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+            throws IOException {
+
+        String userResourcePath = getUserPath() + "?filter=" + filter;
+        HttpGet request = new HttpGet(userResourcePath);
+        request.addHeader(HttpHeaders.AUTHORIZATION, getAuthzHeader());
+        request.addHeader(HttpHeaders.CONTENT_TYPE, "application/json");
+
+        HttpResponse response = client.execute(request);
+        assertEquals(response.getStatusLine().getStatusCode(), 200, "User " +
+                "has not been retrieved successfully");
```
Comment (Abilashini) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java]: https://github.com/wso2/product-is/pull/3601/files#r317469532

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+    private void deleteGroup(String groupId) throws IOException {
+
+        String userResourcePath = getGroupPath() + "/" + groupId;
+        HttpDelete request = new HttpDelete(userResourcePath);
+        request.addHeader(HttpHeaders.AUTHORIZATION, getAuthzHeader());
+        request.addHeader(HttpHeaders.CONTENT_TYPE, "application/json");
+
+        HttpResponse response = client.execute(request);
+        assertEquals(response.getStatusLine().getStatusCode(), 204, "User " +
+                "has not been retrieved successfully");
```
Comment (Abilashini) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java]: https://github.com/wso2/product-is/pull/3601/files#r317469532

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+        EntityUtils.consume(response.getEntity());
+
+        userResourcePath = getGroupPath() + "/" + groupId;
+        HttpGet getRequest = new HttpGet(userResourcePath);
+        getRequest.addHeader(HttpHeaders.AUTHORIZATION, getAuthzHeader());
+        getRequest.addHeader(HttpHeaders.CONTENT_TYPE, "application/json");
+
+        response = client.execute(getRequest);
+        assertEquals(response.getStatusLine().getStatusCode(), 404, "User " +
+                "has not been deleted successfully");
```
Comment (Abilashini) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java]: https://github.com/wso2/product-is/pull/3601/files#r317469532

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+    public SCIM2MultiAttributeUserFilterTestCase(TestUserMode userMode) throws Exception {
+
+        AutomationContext context = new AutomationContext("IDENTITY", userMode);
+        this.adminUsername = context.getContextTenant().getTenantAdmin().getUserName();
+        this.adminPassword = context.getContextTenant().getTenantAdmin().getPassword();
+        this.tenant = context.getContextTenant().getDomain();
+    }
+
+    @DataProvider(name = "SCIM2UserConfigProvider")
+    public static Object[][] sCIM2UserConfigProvider() {
```
Comment (Abilashini) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java]: Maintain the consistency when naming https://github.com/wso2/product-is/pull/3601/files#diff-cbb1ccc927c780601c2258b9b69028b5R109

**Comment by sarubi** [comment]
@Abilashini 
Addressed with https://github.com/wso2/product-is/pull/6381

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java (hunk 2 above)
Comment (sarubi) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java]: Fixed

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java (hunk 8 above)
Comment (sarubi) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java]: Fixed

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+import static org.wso2.identity.integration.test.scim2.SCIM2BaseTestCase.PASSWORD_ATTRIBUTE;
+import static org.wso2.identity.integration.test.scim2.SCIM2BaseTestCase.SCHEMAS_ATTRIBUTE;
+import static org.wso2.identity.integration.test.scim2.SCIM2BaseTestCase.SCIM2_GROUPS_ENDPOINT;
+import static org.wso2.identity.integration.test.scim2.SCIM2BaseTestCase.SCIM2_USERS_ENDPOINT;
+import static org.wso2.identity.integration.test.scim2.SCIM2BaseTestCase.SERVER_URL;
+import static org.wso2.identity.integration.test.scim2.SCIM2BaseTestCase.TYPE_PARAM;
+import static org.wso2.identity.integration.test.scim2.SCIM2BaseTestCase.USER_NAME_ATTRIBUTE;
+import static org.wso2.identity.integration.test.scim2.SCIM2BaseTestCase.VALUE_PARAM;
+
+public class SCIM2MultiAttributeUserFilterTestCase extends ISIntegrationTest {
```
Comment (emswbandara) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java]: add a class level comment

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+
+        assertEquals(response.getStatusLine().getStatusCode(), 201, "User " +
+                "has not been created successfully");
+
+        Object responseObj = JSONValue.parse(EntityUtils.toString(response.getEntity()));
+        EntityUtils.consume(response.getEntity());
+
+        String groupNameFromResponse = ((JSONObject) responseObj).get(SCIM2BaseTestCase.DISPLAY_NAME_ATTRIBUTE)
+                .toString();
+        //TODO: groupsName comes from this is <Userstore_domain>/displayName
```
Comment (emswbandara) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java]: please create a github issue to track this and remove any TODOs in code

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+
+        String groupId = ((JSONObject) responseObj).get(ID_ATTRIBUTE).toString();
+        groupIds.add(groupId);
+        assertNotNull(groupId);
+    }
+
+    @Test(dependsOnMethods = "testCreateGroup")
+    public void testMultiAttributeFilterUser() throws Exception {
+
+        //Validate username and claim filter
```
Comment (emswbandara) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java]: use correct format for comments. Add space after // and fullstop at the end. 
Please check other places as well.

**Comment by sarubi** [comment]
@emswbandara addressed with https://github.com/wso2/product-is/pull/6381

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java (hunk 11 above)
Comment (sarubi) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java]: fixed with https://github.com/wso2/product-is/pull/6381

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java (hunk 10 above)
Comment (sarubi) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java]: created git issue https://github.com/wso2/product-is/issues/6503

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java (hunk 9 above)
Comment (sarubi) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/scim2/SCIM2MultiAttributeUserFilterTestCase.java]: added

**Pull request merged**